### PR TITLE
provide new problem suite matedtrack.epd

### PR DESCRIPTION
**Pull request by robertnurnberg**
The file `matetrack.epd` contains just 26 positions with the side to be mated to move. This makes it hard for engine developers to diagnose and improve their engines' behaviour in such situations. For example, stockfish is known to announce too short mates when it is to be mated, while the mates it announces as the winning side are guaranteed to be correct (so either shortest mate possible or longer).

This draft PR provides the adapted file `matedtrack.epd` that contains many more of the described positions. (As of this moment it contains ~~2978~~ ~~4445~~ ~~6055~~ ~~6058~~ 6465 such positions.)

These positions were created from `matetrack.epd` by using reliable engines (like [Stockfish](https://github.com/official-stockfish/Stockfish), [Huntsman](https://github.com/joergoster/Stockfish-old) and [Ethereal](https://github.com/AndyGrant/Ethereal)) to find a PV for the known best mate value, and then store the fen after the first move with the adjusted best mate. Scripts to do this can be found [here](https://github.com/robertnurnberg/matetrack/tree/mate2pv).

If the maintainer is happy to merge this file into the repo, I can undraft if after a while. (So far I used stockfish and huntsman with 1M nodes to find the best moves, I plan to repeat this with 10M nodes, for example). Otherwise I would ask for this PR to be kept open for a bit, so that developers have a chance to see and download the set of positions.

Edit: Inside my [mate2pv](https://github.com/robertnurnberg/matetrack/tree/mate2pv) branch I have created the file `matetrackpv.epd` with guaranteed PVs for the mates in `matetrack.epd`. I used Ethereal for the mated-in positions. The scripts `pv2mated.py` can then be used to create the file `matedtrack.epd` that is part of this PR. I think long term it may be useful to store the PVs directly on the matetrack repo as well.

### CODE REVIEW

**Comment by robertnurnberg** [comment]
I should have added that the order of the positions in `matedtrack.epd` is the same as in `matetrack.epd`. I.e. they will contain the same FEN where `bm` from `matetrack.epd` was negative or equal to 1, and where I was unable to find a guaranteed PV for a best mate with `bm > 1`. For all other positions `matedtrack.epd` will contain the FEN of the position in `matetrack.epd` plus one ply along a best mating line.

**Comment by robertnurnberg** [comment]
Updated the readme to reflect the additional epd file, and undrafted the PR.

**Comment by robertnurnberg** [comment]
For completeness, I attach the output of `python matecheck.py --epdFile matedtrack.epd` for `sf-dev`:
```
Found mate #-4 (better) for FEN "1R2NnqN/3pPpQ1/b3nPpp/K5Rk/4pPrr/6Pp/4B2P/3N4 b - -".
Found mate #-6 (better) for FEN "r1r3k1/1pp1Rpp1/p1q4p/4B3/6Q1/3B2K1/PPP2PP1/R7 b - -".
Found mate #-8 (better) for FEN "8/2p5/5B2/8/4B3/2K5/pr6/k7 b - -".
Found mate #-44 (better) for FEN "2N5/p4P2/p4RqP/P2krb2/5RNK/2PPBPP1/3P1p2/5Q2 b - -".

Using ./stockfish with 1000000 nodes
Total fens:    6560
Found mates:   4003
Best mates:    3239
Better mates:  5
```

**Comment by robertnurnberg** [comment]
More seriously, the output of `python matecheck.py --epdFile matedtrack.epd --nodes 100` for `sf-dev` gives:
```
Found mate #-1 (better) for FEN "4N2q/3B3p/5p1p/5K1k/7p/7P/8/8 b - -".
Found mate #0 (wrong sign) for FEN "n2n4/1bp5/q2p4/1r1B4/nr1P4/kQ1K4/P2P4/RR6 b - -".
Found mate #-2 (better) for FEN "n2n4/1bp5/q2p4/1r1B4/pr1P4/kQ1K4/P2P4/RR6 b - -".
Found mate #0 (wrong sign) for FEN "6Qk/PP1PP1np/1N1BN2P/7R/4B3/2Q5/P3KP2/6R1 b - -".
Found mate #-7 (better) for FEN "2rq4/3rk3/3bn1Bp/2Np1NP1/8/1p6/1K6/Bn6 b - -".

Using ./stockfish with 100 nodes
Total fens:    6560
Found mates:   24
Best mates:    20
Better mates:  3
Wrong mates:   2
```

so sf seems to return `mate 0` in some instances. See also [discord](https://discord.com/channels/435943710472011776/1032922913499783169/1187695484173832202)

**Comment by robertnurnberg** [comment]
Putting also this back to draft. Needs some more mate corrections from the other PR, and maybe also a general discussion about whether this is the right way forward.

Alternatives are a more thorough revamp of the repo, maybe clearly distinguishing positive and negative mates, and only including those (with mate PV) that we trust.

**Comment by vondele** [comment]
I'm not sure we should really make that differentiation, if we give mate we can find mate sequences longer or equal to the best known mate sequence, if we're being mated we can find mate sequences that are shorter or equal to the best best known. It is also quite logical this property must hold... I think.

**Comment by robertnurnberg** [comment]
If we give mate-in then this is guaranteed to be a mate-in position, but there might be faster mates.

If SF master gives mated-in there is no guarantee this is a lost position. In fact it could be a mate-in position.

So there is some merit in only testing mate-in positions, for as long as SF is not doing something special in mated-in positions.

The idea behind providing a set of many of the latter positions was to offer developers who want to address that aspect a test suite.

But I am now working on providing a list of mating PVs for the test problems in matetrack.epd, then these special positions could be computed on the fly.

**Comment by peregrineshahin** [comment]
> Putting also this back to draft. Needs some more mate corrections from the other PR, and maybe also a general discussion about whether this is the right way forward.
> 
> Alternatives are a more thorough revamp of the repo, maybe clearly distinguishing positive and negative mates, and only including those (with mate PV) that we trust.

Hi, I just want to note again here that an alpha beta engine is not designed to prove mates for opponent, any mated in score raises alpha from -INF, it is impossible to implement or fix the mated in score that this script consider found as better mates for opponent. i.e. current SF master has no problems in mated-in positions, the script assumes that an engine operates as a table-base which will never be the case as long it uses alpha beta and negamax, we are bounded by deep design issues if we think about tweaking Stockfish to this concept.

finding that we are `mated in correct_ply` fast is good and IMO should not be left out from tracking.
if anything should be fixed IMO it is the differentiation between positive and negative scores for finding "better mates".

**Comment by robertnurnberg** [comment]
For completeness, I link [here](https://discord.com/channels/435943710472011776/1032922913499783169/1192767728633794650) to some useful information from @AndyGrant about how also losing mate scores can be guaranteed by an AB engine.

**Comment by robertnurnberg** [comment]
> I'm not sure we should really make that differentiation, if we give mate we can find mate sequences longer or equal to the best known mate sequence, if we're being mated we can find mate sequences that are shorter or equal to the best best known. It is also quite logical this property must hold... I think.

Just on this last point, it can be seen from https://github.com/vondele/matetrack/pull/31#issuecomment-1877441895 that SF currently can also report _longer_ mate scores than the best if it is getting mated. In particular, for 1M nodes it reports `Found mate #-7 (better) for FEN "6k1/2B5/6p1/2K2p2/6n1/8/4n3/2q5 w - -".` for the position `6k1/2B5/6p1/2K2p2/6n1/8/4n3/2q5 w - - bm #-6; Defence; 00:04;`. With 10M nodes it correctly finds `#-6`.

So at present any `mate -X` score given by SF seems to come without _any_ guarantee.

**Comment by AndyGrant** [comment]
I think there is a misunderstanding about the guarantee. What you quoted seems like good behavior to me. Stockfish has found a mate in -7. The guarantee is that Stockfish will not suddenly decide that there is actually a way to delay the mate until -10.

For clarity, a faithful alpha-beta engine ( to my knowledge; and until shown a counter example in a simple graph form ) guarantees that:

1. When reporting MATE in X, there indeed exists a forced mate in X moves. There may also exist a mate in < X moves. However, there should be no way that a perfect opponent can survive for > X moves.
2. When reporting MATED in X, there is indeed a forced mate in X moves, for our opponent, if our opponent is perfect. There may exist a forced mate in < X moves. There however cannot be any move such that our opponent will require > X moves to mate us.

**Comment by peregrineshahin** [comment]
Hi! I don't get what is hard for you to imagine, can we agree that alpha can be negative infinity even though we are not at the first root search iteration?
https://github.com/official-stockfish/Stockfish/pull/4822#issuecomment-1751931070
this reports a mated in 5 while we can delay it to mated in 6
look here https://github.com/official-stockfish/Stockfish/pull/4822#issuecomment-1751931070
Here is the output for this branch
https://github.com/official-stockfish/Stockfish/compare/master...peregrineshahin:Stockfish:candriu-patch
```
$ ./stockfish.exe
Stockfish dev-20240107-nogit by the Stockfish developers (see AUTHORS file)
position fen 6k1/2B5/6p1/2K2p2/6n1/8/4n3/2q5 w - -
go nodes 1000000
info string NNUE evaluation using nn-0000000000a0.nnue
info depth 1 seldepth 1 multipv 1 score cp -854 nodes 7 nps 3500 hashfull 0 tbhits 0 time 2 pv c5d6
info depth 2 seldepth 2 multipv 1 score cp -854 nodes 17 nps 5666 hashfull 0 tbhits 0 time 3 pv c5d6
info depth 3 seldepth 2 multipv 1 score cp -854 nodes 22 nps 5500 hashfull 0 tbhits 0 time 4 pv c5d6
info depth 4 seldepth 2 multipv 1 score cp -893 nodes 64 nps 12800 hashfull 0 tbhits 0 time 5 pv c5b
6
info depth 5 seldepth 3 multipv 1 score cp -865 nodes 131 nps 18714 hashfull 0 tbhits 0 time 7 pv c5
b6 c1e3 b6c6
info depth 6 seldepth 4 multipv 1 score cp -892 nodes 205 nps 20500 hashfull 0 tbhits 0 time 10 pv c
5b6 c1b2 b6a6 b2a2 c7a5
info depth 7 seldepth 6 multipv 1 score cp -866 nodes 290 nps 24166 hashfull 0 tbhits 0 time 12 pv c
5b6 c1b1 b6c6
info depth 8 seldepth 6 multipv 1 score cp -857 nodes 595 nps 33055 hashfull 0 tbhits 0 time 18 pv c
5b6 c1b2 b6a6 b2a1 a6b7 a1h1 b7b8
info depth 9 seldepth 9 multipv 1 score cp -860 nodes 1283 nps 45821 hashfull 0 tbhits 0 time 28 pv
c5b6 c1g1 b6c6
info depth 10 seldepth 9 multipv 1 score cp -870 nodes 2042 nps 49804 hashfull 0 tbhits 0 time 41 pv
 c5b6 c1b2 b6a6 b2a1 a6b7 a1h1 b7a7 e2d4
info depth 11 seldepth 9 multipv 1 score cp -884 nodes 3867 nps 59492 hashfull 1 tbhits 0 time 65 pv
 c5d6 c1d2 d6c6 e2d4 c6b7 d2g2 b7b8
info depth 12 seldepth 12 multipv 1 score cp -907 nodes 8460 nps 63609 hashfull 3 tbhits 0 time 133
pv c5b6 g4f6 b6b7 e2d4 b7c8 c1c6
info depth 13 seldepth 11 multipv 1 score cp -951 nodes 17325 nps 64166 hashfull 6 tbhits 0 time 270
 pv c5b6 e2d4 b6a6 c1c6 c7b6 c6a8 b6a7 a8c8 a6a5
info depth 14 seldepth 16 multipv 1 score cp -978 nodes 49758 nps 73934 hashfull 13 tbhits 0 time 67
3 pv c5b6 e2d4 c7d8 c1c6 b6a7 c6d7 a7a6 d7d8
info depth 15 seldepth 20 multipv 1 score cp -1005 nodes 77750 nps 78062 hashfull 20 tbhits 0 time 9
96 pv c5d6 e2d4 d6d5 c1c6 d5d4 c6c7 d4d3 g4f6 d3d4 f6e4 d4e3 c7b6 e3f4
info depth 16 seldepth 24 multipv 1 score mate -20 nodes 118559 nps 80378 hashfull 33 tbhits 0 time
1475 pv c5d6 e2d4 d6d5 c1c6 d5d4 c6c7 d4d5 c7b6 d5c4 g4f6 c4d3 f6e4 d3c2 g8f7 c2c1 b6c5 c1b2 f5f4 b2
b3
info depth 17 seldepth 19 multipv 1 score mate -20 nodes 120713 nps 80314 hashfull 33 tbhits 0 time
1503 pv c5d6 e2d4 d6d5 c1c6 d5d4 c6c7 d4d5 c7b6 d5c4 g4f6 c4d3 f6e4 d3c2 g8f7 c2c1 b6c5 c1b2 f5f4 b2
b3
info depth 18 seldepth 23 multipv 1 score mate -8 nodes 129183 nps 81094 hashfull 33 tbhits 0 time 1
593 pv c5d6 e2d4 d6d5 c1c7 d5d4 g4f6 d4d3 f6e4 d3d4 c7c5 d4d3 c5c3 d3e2 c3d2 e2f1 d2f2
info depth 19 seldepth 15 multipv 1 score mate -7 nodes 134368 nps 81287 hashfull 35 tbhits 0 time 1
653 pv c5d6 e2d4 d6d5 c1c7 d5d4 g4f6 d4d3 f6e4 d3d4 c7c2 d4d5 c2c5 d5e6 c5d6
info depth 20 seldepth 15 multipv 1 score mate -7 nodes 139444 nps 80930 hashfull 36 tbhits 0 time 1
723 pv c5d6 e2d4 d6d5 c1c7 d5d4 g4f6 d4d3 f6e4 d3d4 c7c2 d4d5 c2c5 d5e6 c5d6
info depth 21 seldepth 15 multipv 1 score mate -7 nodes 149006 nps 81424 hashfull 38 tbhits 0 time 1
830 pv c5d6 e2d4 d6d5 c1c7 d5d4 g4f6 d4d3 f6e4 d3d4 c7c2 d4d5 c2c5 d5e6 c5d6
info depth 22 seldepth 15 multipv 1 score mate -7 nodes 160816 nps 81715 hashfull 42 tbhits 0 time 1
968 pv c5d6 e2d4 d6d5 c1c7 d5d4 g4f6 d4d3 f6e4 d3d4 c7c2 d4d5 c2c5 d5e6 c5d6
info depth 23 seldepth 15 multipv 1 score mate -7 nodes 183462 nps 81683 hashfull 47 tbhits 0 time 2
246 pv c5d6 e2d4 d6d5 c1c7 d5d4 g4f6 d4d3 f6e4 d3d4 c7c2 d4d5 c2c5 d5e6 c5d6
info depth 24 seldepth 15 multipv 1 score mate -7 nodes 226094 nps 81037 hashfull 59 tbhits 0 time 2
790 pv c5d6 e2d4 d6d5 c1c7 d5d4 g4f6 d4d3 f6e4 d3d4 c7c2 d4d5 c2c5 d5e6 c5d6
info depth 25 currmove c5b4 currmovenumber 2
info depth 25 currmove c5b6 currmovenumber 3
info depth 25 currmove c5d5 currmovenumber 4
info depth 25 currmove c5b5 currmovenumber 5
info depth 25 seldepth 15 multipv 1 score mate -7 nodes 281312 nps 80628 hashfull 73 tbhits 0 time 3
489 pv c5d6 e2d4 d6d5 c1c7 d5d4 g4f6 d4d3 f6e4 d3d4 c7c2 d4d5 c2c5 d5e6 c5d6
info depth 26 currmove c5d6 currmovenumber 1
info depth 26 currmove c5d5 currmovenumber 2
info depth 26 currmove c5b4 currmovenumber 3
info depth 26 currmove c5b5 currmovenumber 4
info depth 26 currmove c5b6 currmovenumber 5
info depth 26 seldepth 15 multipv 1 score mate -7 nodes 349550 nps 80190 hashfull 92 tbhits 0 time 4
359 pv c5d6 e2d4 d6d5 c1c7 d5d4 g4f6 d4d3 f6e4 d3d4 c7c2 d4d5 c2c5 d5e6 c5d6
info depth 27 currmove c5d6 currmovenumber 1
info depth 27 currmove c5d5 currmovenumber 2
info depth 27 currmove c5b4 currmovenumber 3
info depth 27 currmove c5b6 currmovenumber 4
info depth 27 currmove c5b5 currmovenumber 5
info depth 27 seldepth 15 multipv 1 score mate -7 nodes 437921 nps 80470 hashfull 122 tbhits 0 time
5442 pv c5d6 e2d4 d6d5 c1c7 d5d4 g4f6 d4d3 f6e4 d3d4 c7c2 d4d5 c2c5 d5e6 c5d6
info depth 28 currmove c5d6 currmovenumber 1
info depth 28 currmove c5b4 currmovenumber 2
info depth 28 currmove c5d5 currmovenumber 3
info depth 28 currmove c5b6 currmovenumber 4
info depth 28 currmove c5b5 currmovenumber 5
info depth 28 seldepth 15 multipv 1 score mate -7 nodes 562973 nps 79933 hashfull 160 tbhits 0 time
7043 pv c5d6 e2d4 d6d5 c1c7 d5d4 g4f6 d4d3 f6e4 d3d4 c7c2 d4d5 c2c5 d5e6 c5d6
info depth 29 currmove c5d6 currmovenumber 1
info depth 29 currmove c5b4 currmovenumber 2
info depth 29 currmove c5d5 currmovenumber 3
info depth 29 currmove c5b6 currmovenumber 4
info depth 29 currmove c5b5 currmovenumber 5
info depth 29 seldepth 15 multipv 1 score mate -7 nodes 729986 nps 79372 hashfull 207 tbhits 0 time
9197 pv c5d6 e2d4 d6d5 c1c7 d5d4 g4f6 d4d3 f6e4 d3d4 c7c2 d4d5 c2c5 d5e6 c5d6
info depth 30 currmove c5d6 currmovenumber 1
value: -31990
alpha: -32001
beta: 32001
depth: 30
Root Delta: 64002
```

Now tell me what's hard to imagine about this?
We have raised alpha of negative infinity and before searching other root moves we updated root score, so what's hard?

**Comment by AndyGrant** [comment]
We've had this conversation already. No value in us talking to each other any further.

**Comment by robertnurnberg** [comment]
> I think there is a misunderstanding about the guarantee. What you quoted seems like good behavior to me. Stockfish has found a mate in -7. The guarantee is that Stockfish will not suddenly decide that there is actually a way to delay the mate until -10.

The misunderstanding developed due to the peculiar behaviour of Stockfish when reporting `mate -X` scores. I completely agree with your points 1. and 2. I think 1. is noncontroversial, and probably what almost every chess engine implements or tries to implement. I also agree that 2. is what should ideally be done, and that that is the behaviour users would and should expect.

Unfortunately, it has been known for some time now that current Stockfish violates 2. I.e. it may report `mate -5` in a position that is `mate -6`, and it itself will find the correct `mate -6` with more nodes/at higher depth. It is only because of that peculiar behaviour that some of us believed that Stockfish may instead provide a subjective currently best line for itself, which would include the case of reporting `mate -5` in a `mate -6` position. I think this is what vondele meant with

> if we give mate we can find mate sequences longer or equal to the best known mate sequence, if we're being mated we can find mate sequences that are shorter or equal to the best best known

in https://github.com/vondele/matetrack/pull/25#issuecomment-1869610000, and which is what led to my ill-fated PR https://github.com/vondele/matetrack/pull/31. 

In summary, it seems that current Stockfish, for a `mate -6` position, may report `mate -5`, `mate -6` or `mate -7`. More generally, it may report `mate -X` even in a drawn or in a won position, although I have no example position ready to prove this at the moment.

**Comment by robertnurnberg** [comment]
Please see https://github.com/robertnurnberg/matetrack/tree/ethereal for a way to test engines for mated-in problems. Ethereal passes with flying colours, while Stockfish fails. See also [discord](https://discord.com/channels/435943710472011776/1032922913499783169/1193566568588791919).

**Comment by joergoster** [comment]
@robertnurnberg I just pushed some commits to my old [huntsman](https://github.com/joergoster/Stockfish-old/tree/huntsman) branch. If you're able to compile yourself feel free to give it a try.

It should avoid output of incomplete searches at all, at the small cost of some decrease in performance when running with a nodes or movetime limit.

**Comment by robertnurnberg** [comment]
> @robertnurnberg I just pushed some commits to my old [huntsman](https://github.com/joergoster/Stockfish-old/tree/huntsman) branch. If you're able to compile yourself feel free to give it a try.
> 
> It should avoid output of incomplete searches at all, at the small cost of some decrease in performance when running with a nodes or movetime limit.

Very nice, thanks. I can confirm that your latest huntsman version passes `test_engine.sh` from https://github.com/robertnurnberg/matetrack/tree/ethereal in the same way that Ethereal does, while Huntsman 1.0 fails with a lot of false mates.

**Comment by robertnurnberg** [comment]
Maybe you could try to apply your fix(es) to SF itself, at least for the wrong `mate -X` scores. It should be Elo neutral and a big improvement for users in general. @joergoster

**Comment by joergoster** [comment]
@robertnurnberg Thank you for running the test. Greatly appreciated!

> Maybe you could try to apply your fix(es) to SF itself, at least for the wrong mate -X scores. It should be Elo neutral and a big improvement for users in general.

I'll leave that to others.

**Comment by joergoster** [comment]
@robertnurnberg Here the results of current huntsman-dev with 1000, 10,000 and 100,000 nodes.

```
Using ./huntsman with --nodes 1000
Total fens:    6560
Found mates:   46
Best mates:    35
Complete PVs:  46/46 (100.0%)
```

```
Using ./huntsman with --nodes 10000
Total fens:    6560
Found mates:   435
Best mates:    356
Complete PVs:  435/435 (100.0%)
```

```
Using ./huntsman with --nodes 100000
Total fens:    6560
Found mates:   2485
Best mates:    1792
Complete PVs:  2485/2485 (100.0%)
```
Tests were run with current matetrack-dev.
Are there any further tests you'd recommend?
Note: Current dev-version is not available online atm.

**Comment by robertnurnberg** [comment]
Very nice @joergoster . I especially like the 100% PV completion.

You could also run `python matecheck.py --epdFile matedtrack.epd --engine ./huntsman` on this updated PR, to test the mated-in positions. (Unless the output you state is already for this file, not clear.) If the runs you show are already for the `matedtrack.epd` file, you could do the same for `python matecheck.py --engine ./huntsman` on this PR (as it includes some corrected mates compared to master).

Lastly, the fix https://github.com/official-stockfish/Stockfish/pull/4976 under rare circumstances suffers from too short mates with high threads. So one more check for you would be `python matecheck.py --epdFile matedtrack.epd --engine ./huntsman --threads 8`.

**Comment by joergoster** [comment]
Here the output of

> python3 matecheck.py --epdFile matedtrack.epd --engine ./huntsman --nodes 100000 --concurrency 2

```
Using ./huntsman with --nodes 100000
Total fens:    6560
Found mates:   3577
Best mates:    2806
Complete PVs:  3577/3577 (100.0%)
```

**Comment by joergoster** [comment]
Trying to run the following command to also test the SMP part, I get an error:

> python3 matecheck.py --epdFile matedtrack.epd --engine ./huntsman --nodes 200000 --concurrency 1 --threads 4
> Loading FENs...
> 6560 FENs loaded...
> Traceback (most recent call last):
>   File "/home/joster/Testsets/Matetrack/matetrack-master-2024-01/matecheck.py", line 126, in <module>
>     fw_ratio = numfen // (4 * workers)
> ZeroDivisionError: integer division or modulo by zero

**Comment by joergoster** [comment]
Now 2 SMP results ...

> python3 matecheck.py --engine ./huntsman --nodes 200000 --concurrency 4 --threads 4

```
Using ./huntsman with --nodes 200000 --threads 4
Total fens:    6560
Found mates:   2928
Best mates:    2159
Complete PVs:  2928/2928 (100.0%)
```
and

> python3 matecheck.py --epdFile matedtrack.epd --engine ./huntsman --nodes 200000 --concurrency 4 --threads 4

```
Using ./huntsman with --nodes 200000 --threads 4
Total fens:    6560
Found mates:   4077
Best mates:    3236
Complete PVs:  4077/4077 (100.0%)
```
Looks good!

**Comment by robertnurnberg** [comment]
So what started as a little experiment quickly grew into something bigger. To keep track, I have started the repo https://github.com/robertnurnberg/matetools with the scripts I used to obtain the mating PVs.

@vondele At some stage it might be a good idea to keep track of the mating PVs also in this repo. For example, the mating PVs for the positions in `matetrack.epd` can be used to create an almost arbitrary number of new mate puzzles (with shorter mates, of course). One application is the file `matedtrack.epd`.

Before I close this and open a new PR, what do you think the best place for storing the PVs is: inside `matetrack.epd` itself, or as a separate file `matetrackpv.epd`? See my repo for what the latter file looks like. The same then goes for `matedtrack.epd`. My current thinking is to maybe have three files in the matetrack repo:  `matetrack.epd` as now, plus `matetrackpv.epd` and `matedtrack.epd` as on my repo. One could refer to my repo on how to create these, but not strictly necessary.

**Comment by vondele** [comment]
I think it makes sense to have these files, it is useful to have a full solution line where available.

**Comment by robertnurnberg** [comment]
> I think it makes sense to have these files, it is useful to have a full solution line where available.

Agreed. Do you prefer to have the PV inside the `matetrack.epd` itself, or in a separate file?

**Comment by vondele** [comment]
no strong preference, probably a separate file? Depends on what you find easiest to use.

**Comment by robertnurnberg** [comment]
Ok, thanks. Separate file it is: for continuity and to not overburden the original file with info.

**Comment by robertnurnberg** [comment]
Closing in favour of https://github.com/vondele/matetrack/pull/41.
